processing/src/main/java/org/apache/druid/query/aggregation/firstlast/FirstLastUtils.java

```diff
@@ -0,0 +1,158 @@
+ */
+
+package org.apache.druid.query.aggregation.firstlast;
+
+import org.apache.druid.query.aggregation.SerializablePairLongDouble;
+import org.apache.druid.query.aggregation.SerializablePairLongFloat;
+import org.apache.druid.query.aggregation.SerializablePairLongLong;
+import org.apache.druid.segment.BaseObjectColumnValueSelector;
+import org.apache.druid.segment.DimensionHandlerUtils;
+import org.apache.druid.segment.NilColumnValueSelector;
+import org.apache.druid.segment.column.ColumnCapabilities;
+import org.apache.druid.segment.column.ValueType;
+
+import javax.annotation.Nullable;
+
+public class FirstLastUtils
+{
+
+  /**
+   * Returns whether a given value selector *might* contain object assignable from pairClass (SerializablePairLong*).
+   */
+  public static boolean selectorNeedsFoldCheck(
+      final BaseObjectColumnValueSelector<?> valueSelector,
+      @Nullable final ColumnCapabilities valueSelectorCapabilities,
+      Class pairClass
+  )
+  {
+    if (valueSelectorCapabilities != null && !valueSelectorCapabilities.is(ValueType.COMPLEX)) {
+      // Known, non-complex type.
+      return false;
+    }
+
+    if (valueSelector instanceof NilColumnValueSelector) {
+      // Nil column, definitely no SerializablePairLongObject.
+      return false;
+    }
+
+    // Check if the selector class could possibly be of pairClass* (either a superclass or subclass).
+    final Class<?> clazz = valueSelector.classOfObject();
+    return clazz.isAssignableFrom(pairClass)
+           || pairClass.isAssignableFrom(clazz);
+  }
+
+  @Nullable
+  public static SerializablePairLongDouble readDoublePairFromVectorSelectors(
+      @Nullable boolean[] timeNullityVector,
+      long[] timeVector,
+      Object[] objectVector,
+      int index
+  )
+  {
+    final long time;
+    final Double value;
+
+    final Object object = objectVector[index];
+
+    if (object instanceof SerializablePairLongDouble) {
+      // We got a folded object, ignore timeSelector completely, the object has all the info it requires
+      final SerializablePairLongDouble pair = (SerializablePairLongDouble) object;
+      // if time == null, don't aggregate
+      if (pair.lhs == null) {
+        return null;
+      }
+      time = pair.lhs;
+      value = pair.rhs;
+    } else {
+      if (timeNullityVector != null && timeNullityVector[index]) {
+        // Donot aggregate pairs where time is unknown
+        return null;
+      }
+      time = timeVector[index];
+      value = DimensionHandlerUtils.convertObjectToDouble(object);
+    }
+    return new SerializablePairLongDouble(time, value);
+  }
+
+  @Nullable
+  public static SerializablePairLongFloat readFloatPairFromVectorSelectors(
+      @Nullable boolean[] timeNullityVector,
+      long[] timeVector,
+      Object[] objectVector,
+      int index
+  )
+  {
+    final long time;
+    final Float value;
+
+    final Object object = objectVector[index];
+
+    if (object instanceof SerializablePairLongFloat) {
```

Comment:
I find it odd that if the value is a `Pair` ; `timeNullityVector` is not considered at all...

it seems to be it was working like this even before ; so it could be ok....

Reply:
If there's a Pair, then we ignore everything supplied by the timeSelector. All the information about the time is encoded in that pair. A pair could be something like: {lhs: time1, rhs: value1}, and the timeNullity vector could be [true]. We **need** to ignore whatever is present in the timeVector, and proceed with the information present inside the pair.

Here's a PR https://github.com/apache/druid/pull/15939, where in the non-vectorised versions, this flow was broken - we checked the time's nullity first and then considered the pairs. That led to incorrect results